### PR TITLE
[JENKINS-73768] Handle base=derived case in Util#isOverridden

### PR DESCRIPTION
**Pull request by daniel-beck**
See [JENKINS-73768](https://issues.jenkins.io/browse/JENKINS-73768).

It looks like https://github.com/jenkinsci/jenkins/pull/4833 does not properly consider the case where base == derived. Somehow the existing test coverage is insufficient, did not dig further into this yet.

I encountered this when looking at the system log and being surprised about exception output using "Caused by" rather than "Causes" when [what was thrown was a raw `Throwable`](https://github.com/jenkinsci/jenkins/blob/d5271c8d2f38a2ba8b1f9d8fcbb66070e5e5333d/core/src/main/java/hudson/Functions.java#L1813-L1818) in test code -- hence the choice of types in the test.

An alternative solution could be to change the last line to use `equals` instead of `!=`, but that seems more likely to have unintended consequences.

CC @Zastai in case you're still around/interested.

### Testing done

None.

### Proposed changelog entries

- human-readable text

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

**Comment by Zastai** [comment]
LGTM.

I suppose the second test assert is not 100% future proof (there's no guarantee that `Exception` will never override `printStackTrace()`). The test class could declare a nested private static class that extends `Throwable` for that purpose; not sure it's necessarily worth the effort though.

File: core/src/main/java/hudson/Util.java
Comment (daniel-beck) [inline core/src/main/java/hudson/Util.java:1529]:
```suggestion
            // If base and derived are the same type, the method is not overridden by definition
```

**Comment by zbynek** [comment]
Looks like the method was copied to https://github.com/jenkinsci/stapler/blob/c4f6b569053629f8ede0e176bce6538a2450585f/core/src/main/java/org/kohsuke/stapler/ReflectionUtils.java#L102 and both methods are called by core. Would it make sense to also apply the fix there? Or maybe deduplicate the methods?

**Comment by daniel-beck** [comment]
@zbynek 

> Would it make sense to also apply the fix there?

Probably. Note that the problem only matters if you have a caller passing types that are not hardcoded, not the case in Stapler AFAICT.

**Review by timja — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
